From 4e5088a827f36d1ea6da4c59af2c5ad361fa6b47 Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Wed, 8 Oct 2025 16:06:50 +0200
Subject: [PATCH] ruby: disable yjit on cross

Fix `nix build -L nixpkgs#pkgsCross.aarch64-multiplatform.ruby` failing
with:
```
ruby-aarch64-unknown-linux-gnu> partial linking yjit/target/release/libyjit.a into yjit/target/release/libyjit.o
ruby-aarch64-unknown-linux-gnu> /nix/store/fxzmz6h2m0l74kxglsl5q4zcncw19035-aarch64-unknown-linux-gnu-binutils-2.44/bin/aarch64-unknown-linux-gnu-ld: yjit/target/release/libyjit.a(yjit.1fllkjkj80qowi7xzzkaw1xy8.rcgu.o): Relocations in generic ELF (EM: 62)
ruby-aarch64-unknown-linux-gnu> /nix/store/fxzmz6h2m0l74kxglsl5q4zcncw19035-aarch64-unknown-linux-gnu-binutils-2.44/bin/aarch64-unknown-linux-gnu-ld: yjit/target/release/libyjit.a(yjit.1fllkjkj80qowi7xzzkaw1xy8.rcgu.o): Relocations in generic ELF (EM: 62)
ruby-aarch64-unknown-linux-gnu> /nix/store/fxzmz6h2m0l74kxglsl5q4zcncw19035-aarch64-unknown-linux-gnu-binutils-2.44/bin/aarch64-unknown-linux-gnu-ld: yjit/target/release/libyjit.a(yjit.1fllkjkj80qowi7xzzkaw1xy8.rcgu.o): error adding symbols: file in wrong format
```

cross compilation seems unsupported for the rust parts.
---
 pkgs/development/interpreters/ruby/default.nix | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/pkgs/development/interpreters/ruby/default.nix b/pkgs/development/interpreters/ruby/default.nix
index a77ea35bfd3280..de51eb63f900c7 100644
--- a/pkgs/development/interpreters/ruby/default.nix
+++ b/pkgs/development/interpreters/ruby/default.nix
@@ -59,9 +59,11 @@ let
       ver = version;
       atLeast31 = lib.versionAtLeast ver.majMin "3.1";
       atLeast32 = lib.versionAtLeast ver.majMin "3.2";
+      isCross = (stdenv.buildPlatform != stdenv.hostPlatform);
       # https://github.com/ruby/ruby/blob/v3_2_2/yjit.h#L21
       yjitSupported =
         atLeast32
+        && !isCross
         && (
           stdenv.hostPlatform.isx86_64 || (!stdenv.hostPlatform.isWindows && stdenv.hostPlatform.isAarch64)
         );
